1-Draft/RFCNNNN-Automatic-Resource-Management.md

```diff
@@ -0,0 +1,197 @@
+---
+RFC: RFC<four digit unique incrementing number assigned by Committee, this shall be left blank by the author>
+Author: Ilya Sazonov
+Status: Draft
+SupercededBy: <link to another RFC>
+Version: 1.0
+Area: Engine
+Comments Due: 7/1/2020
+Plan to implement: Maybe
+---
+
+# Automatic resource management
+
+## Rationale
+
+Currently PowerShell has limited capabilities compared with C# to excecute a mandatory code.
+
+PowerShell users can benefit from `finally` block of `try - catch` statement. It is very limited experience because resources in PowerShell can be distributed across scopes, cmdlets, pipelines and runspaces.
+
+If an user created a temporary file one should worry about deleting it, while PowerShell could register a mandatory code to delete this file and execute it automatically at the right time.
+
+PowerShell makes many smart things for users and could offer smart resource management.
```

Comment:
I think this phrase is really the crux of the proposal: Should PowerShell attempt to do automatic resource management separately from the CLR?

I think any wide-ranging, possibly destructive feature like this can't just be something based on heuristics but must instead be based on some systematic analysis or formalism. When the garbage collector collects an object, it's based on a proof that the object is no longer in use computed by crawling the heap. When a compiler optimises an unreachable statement out, it's based on a proof that the refinement type of the condition around that statement is bottom (or something similar).

PowerShell has no such established systematic analysis, and I think adding one now would be a dramatic change to the language's semantics. It would be somewhat like adding a garbage collector to C++ today.

Instead, .NET provides a way to integrate resource disposal into their existing system for determining whether a resource is no longer in use (the GC): Implementing `Object.Finalize()` to call `Dispose(false)`.

.NET (and thus PowerShell) already pays a high-price for garbage collection and the analysis thereof, and I think we should position ourselves to take advantage of that.

I also worry about telling PowerShell users that they should feel free to open or acquire resources without any regard for their lifetime and that PowerShell will magically clean it up. With power comes responsibility, and we should aim to favour reliable semantics over magical ergonomics in every circumstance. Rather than try to do things for users based on what we imagine they want, we should give them easy opportunities to declare or configure what they want.